File: backend/routes/vehicles.py
from flask import Blueprint, request, jsonify, current_app
from flask_jwt_extended import jwt_required, get_jwt_identity, get_jwt
import datetime
import os
from bson.objectid import ObjectId
from werkzeug.utils import secure_filename
from dateutil import parser

vehicles_bp = Blueprint("vehicles", __name__)

def get_db():
    from app import db
    return db

@vehicles_bp.route("/add", methods=["POST"])
@jwt_required()
def add_vehicle():
    data = request.form
    
    # Core/Identity
    vehicle_number = data.get("vehicleNumber")
    model = data.get("model")
    year = data.get("year")
    fuel_type = data.get("fuelType")
    brand = data.get("brand")
    variant = data.get("variant")
    vehicle_type = data.get("vehicleType")
    color = data.get("color")
    registration_date = data.get("registrationDate")
    registered_rto = data.get("registeredRTO")
    
    # Ownership
    owner_name = data.get("ownerName")
    phone = data.get("phone")
    ownership_count = data.get("ownershipCount")
    purchase_date = data.get("purchaseDate")
    purchase_price = data.get("purchasePrice")
    
    # Legal
    insurance_provider = data.get("insuranceProvider")
    insurance_expiry = data.get("insuranceExpiry")
    puc_expiry = data.get("pucExpiry")
    rc_expiry = data.get("rcExpiry")
    road_tax_valid_till = data.get("roadTaxValidTill")
    fitness_expiry = data.get("fitnessExpiry")
    
    # Verification
    chassis_number = data.get("chassisNumber")
    engine_number = data.get("engineNumber")
    
    # Usage
    current_odometer_km = data.get("currentOdometerKm")
    average_monthly_km = data.get("averageMonthlyKm", 0)
    
    if not all([vehicle_number, model, year, fuel_type]):
        return jsonify({"msg": "Vehicle number, model, year, and fuel type are required"}), 400
        
    # Validations
    try:
        if year:
            year_int = int(year)
            current_year = datetime.datetime.now().year
            if year_int < 1900 or year_int > current_year + 1:
                return jsonify({"msg": "Manufacturing year must be a valid year"}), 400
        
        # Expiry Date validations
        now = datetime.datetime.now()
        for date_str, name in [
            (insurance_expiry, "Insurance Expiry"), 
            (puc_expiry, "PUC Expiry"), 
            (rc_expiry, "RC Expiry"), 
            (road_tax_valid_till, "Road Tax Valid Till"), 
            (fitness_expiry, "Fitness Expiry")
        ]:
            if date_str:
                parsed_date = parser.parse(date_str)
                # Compare only dates
                if parsed_date.date() < now.date():
                    return jsonify({"msg": f"{name} must be a future date"}), 400
    except ValueError:
        return jsonify({"msg": "Invalid date or number format"}), 400
        
    db = get_db()
    vehicles_collection = db["vehicles"]
    
    # Check if vehicle already exists
    if vehicles_collection.find_one({"vehicleNumber": vehicle_number}):
        return jsonify({"msg": "Vehicle with this number already exists"}), 400
        
    owner_id = get_jwt_identity()
    
    # Handle File Upload
    rc_book_filename = None
    if 'rcBook' in request.files:
        file = request.files['rcBook']
        if file and file.filename != '':
            filename = secure_filename(file.filename)
            # Create a unique filename using timestamp
            unique_filename = f"{datetime.datetime.now().strftime('%Y%m%d%H%M%S')}_{filename}"
            filepath = os.path.join(current_app.config['UPLOAD_FOLDER'], unique_filename)
            file.save(filepath)
            rc_book_filename = unique_filename
    
    new_vehicle = {
        # Identity
        "vehicleNumber": vehicle_number,
        "brand": brand,
        "model": model,
        "variant": variant,
        "vehicleType": vehicle_type,
        "fuelType": fuel_type,
        "color": color,
        "manufacturingYear": year_int if year else None,
        "registrationDate": registration_date,
        "registeredRTO": registered_rto,
        
        # Ownership
        "ownerName": owner_name,
        "phone": phone,
        "ownershipCount": int(ownership_count) if ownership_count else 1,
        "purchaseDate": purchase_date,
        "purchasePrice": float(purchase_price) if purchase_price else None,
        
        # Legal
        "insuranceProvider": insurance_provider,
        "insuranceExpiry": insurance_expiry,
        "pucExpiry": puc_expiry,
        "rcExpiry": rc_expiry,
        "roadTaxValidTill": road_tax_valid_till,
        "fitnessExpiry": fitness_expiry,
        
        # Verification
        "chassisNumber": chassis_number,
        "engineNumber": engine_number,
        "rcBookUrl": f"/uploads/{rc_book_filename}" if rc_book_filename else None,
        
        # Usage
        "currentOdometerKm": int(current_odometer_km) if current_odometer_km else 0,
        "averageMonthlyKm": int(average_monthly_km) if average_monthly_km else 0,
        
        # Audit & Data Integrity
        "ownerId": owner_id,
        "createdBy": owner_id,
        "role": get_jwt().get("role", "Vehicle Owner"),
        "verificationStatus": "Pending",
        "isArchived": False,
        "createdAt": datetime.datetime.utcnow()
    }
    
    try:
        vehicles_collection.insert_one(new_vehicle)
        return jsonify({"msg": "Vehicle added successfully"}), 201
    except Exception as e:
        return jsonify({"msg": "Error adding vehicle", "error": str(e)}), 500


@vehicles_bp.route("/myvehicles", methods=["GET"])
@jwt_required()
def get_my_vehicles():
    db = get_db()
    vehicles_collection = db["vehicles"]
    
    owner_id = get_jwt_identity()
    
    # Filter out soft-deleted (archived) vehicles
    vehicles_cursor = vehicles_collection.find({"ownerId": owner_id, "isArchived": {"$ne": True}})
    vehicles = []
    
    for v in vehicles_cursor:
        vehicles.append({
            "id": str(v["_id"]),
            "vehicleNumber": v.get("vehicleNumber"),
            "brand": v.get("brand"),
            "model": v.get("model"),
            "variant": v.get("variant"),
            "vehicleType": v.get("vehicleType"),
            "year": v.get("manufacturingYear") or v.get("year"), # Backward compatibility
            "fuelType": v.get("fuelType"),
            "color": v.get("color"),
            "registrationDate": v.get("registrationDate"),
            "registeredRTO": v.get("registeredRTO"),
            "ownerName": v.get("ownerName"),
            "phone": v.get("phone"),
            "ownershipCount": v.get("ownershipCount"),
            "purchaseDate": v.get("purchaseDate"),
            "purchasePrice": v.get("purchasePrice"),
            "insuranceProvider": v.get("insuranceProvider"),
            "insuranceExpiry": v.get("insuranceExpiry"),
            "pucExpiry": v.get("pucExpiry"),
            "rcExpiry": v.get("rcExpiry"),
            "roadTaxValidTill": v.get("roadTaxValidTill"),
            "fitnessExpiry": v.get("fitnessExpiry"),
            "chassisNumber": v.get("chassisNumber"),
            "engineNumber": v.get("engineNumber"),
            "currentOdometerKm": v.get("currentOdometerKm"),
            "averageMonthlyKm": v.get("averageMonthlyKm"),
            "ownerId": v.get("ownerId"),
            "rcBookUrl": v.get("rcBookUrl"),
            "verificationStatus": v.get("verificationStatus", "Pending"),
            "isArchived": v.get("isArchived", False),
            "createdAt": v.get("createdAt").isoformat() if v.get("createdAt") else None
        })
        
    return jsonify(vehicles), 200

@vehicles_bp.route("/<vehicle_id>", methods=["PUT"])
@jwt_required()
def edit_vehicle(vehicle_id):
    try:
        data = request.form
        # Extract all fields similar to add_vehicle
        update_data = {}
        for key in ["vehicleNumber", "brand", "model", "variant", "vehicleType", 
                    "fuelType", "color", "registrationDate", "registeredRTO", 
                    "ownerName", "phone", "purchaseDate", "insuranceProvider", 
                    "insuranceExpiry", "pucExpiry", "rcExpiry", "roadTaxValidTill", 
                    "fitnessExpiry", "chassisNumber", "engineNumber"]:
            if key in data:
                update_data[key] = data.get(key)
                
        # Handle int/float specific fields
        if "year" in data and data.get("year"): update_data["manufacturingYear"] = int(data.get("year"))
        if "ownershipCount" in data and data.get("ownershipCount"): update_data["ownershipCount"] = int(data.get("ownershipCount"))
        if "purchasePrice" in data and data.get("purchasePrice"): update_data["purchasePrice"] = float(data.get("purchasePrice"))
        if "currentOdometerKm" in data and data.get("currentOdometerKm"): update_data["currentOdometerKm"] = int(data.get("currentOdometerKm"))
        if "averageMonthlyKm" in data and data.get("averageMonthlyKm"): update_data["averageMonthlyKm"] = int(data.get("averageMonthlyKm"))

        
        db = get_db()
        vehicles_collection = db["vehicles"]
        owner_id = get_jwt_identity()
        
        # Verify ownership and not archived
        vehicle = vehicles_collection.find_one({
            "_id": ObjectId(vehicle_id), 
            "ownerId": owner_id,
            "isArchived": {"$ne": True}
        })
        if not vehicle:
            return jsonify({"msg": "Vehicle not found or unauthorized"}), 404
            
        vehicle_number = update_data.get("vehicleNumber")
        
        # Check if new vehicle number conflicts with another vehicle
        if vehicle_number and vehicle_number != vehicle.get("vehicleNumber"):
            if vehicles_collection.find_one({"vehicleNumber": vehicle_number}):
                return jsonify({"msg": "Vehicle with this number already exists"}), 400
        
        # Handle File Upload for RC Book Update
        if 'rcBook' in request.files:
            file = request.files['rcBook']
            if file and file.filename != '':
                # Delete old file if exists
                old_rc_url = vehicle.get("rcBookUrl")
                if old_rc_url:
                    old_filename = old_rc_url.split("/")[-1]
                    old_path = os.path.join(current_app.config['UPLOAD_FOLDER'], old_filename)
                    if os.path.exists(old_path):
                        os.remove(old_path)
                
                filename = secure_filename(file.filename)
                unique_filename = f"{datetime.datetime.now().strftime('%Y%m%d%H%M%S')}_{filename}"
                filepath = os.path.join(current_app.config['UPLOAD_FOLDER'], unique_filename)
                file.save(filepath)
                update_data["rcBookUrl"] = f"/uploads/{unique_filename}"
                
        if update_data:
            vehicles_collection.update_one(
                {"_id": ObjectId(vehicle_id)},
                {"$set": update_data}
            )
            
        return jsonify({"msg": "Vehicle updated successfully"}), 200
    except Exception as e:
        return jsonify({"msg": "Error updating vehicle", "error": str(e)}), 500

@vehicles_bp.route("/<vehicle_id>", methods=["DELETE"])
@jwt_required()
def delete_vehicle(vehicle_id):
    try:
        db = get_db()
        vehicles_collection = db["vehicles"]
        services_collection = db["services"]
        owner_id = get_jwt_identity()
        
        # Verify ownership
        vehicle = vehicles_collection.find_one({"_id": ObjectId(vehicle_id), "ownerId": owner_id})
        if not vehicle:
            return jsonify({"msg": "Vehicle not found or unauthorized"}), 404
            
        # Soft delete related services
        services_collection.update_many(
            {"vehicleId": vehicle_id},
            {"$set": {"isArchived": True}}
        )
        
        # Soft delete vehicle
        vehicles_collection.update_one(
            {"_id": ObjectId(vehicle_id)},
            {"$set": {"isArchived": True}}
        )
        
        return jsonify({"msg": "Vehicle archived successfully"}), 200
    except Exception as e:
        return jsonify({"msg": "Error archiving vehicle", "error": str(e)}), 500

File: backend/src/routes/vehicles.js
const express = require('express');
const { ObjectId } = require('mongodb');

const { getDb } = require('../db');
const { requireAuth } = require('../middleware/auth');
const { upload, removeUploadByUrl } = require('../utils/uploads');
const { parseAnyDate, isPastDateOnly } = require('../utils/dates');

const router = express.Router();

router.post('/add', requireAuth, upload.single('rcBook'), async (req, res) => {
  const data = req.body || {};

  const vehicleNumber = data.vehicleNumber;
  const model = data.model;
  const year = data.year;
  const fuelType = data.fuelType;

  if (!vehicleNumber || !model || !year || !fuelType) {
    return res.status(400).json({ msg: 'Vehicle number, model, year, and fuel type are required' });
  }

  // Validations
  let yearInt;
  try {
    yearInt = parseInt(year, 10);
    const currentYear = new Date().getFullYear();
    if (Number.isNaN(yearInt) || yearInt < 1900 || yearInt > currentYear + 1) {
      return res.status(400).json({ msg: 'Manufacturing year must be a valid year' });
    }

    const now = new Date();
    const nowUtc = require('luxon').DateTime.fromJSDate(now, { zone: 'utc' });
    const toCheck = [
      { value: data.insuranceExpiry, name: 'Insurance Expiry' },
      { value: data.pucExpiry, name: 'PUC Expiry' },
      { value: data.rcExpiry, name: 'RC Expiry' },
      { value: data.roadTaxValidTill, name: 'Road Tax Valid Till' },
      { value: data.fitnessExpiry, name: 'Fitness Expiry' }
    ];
    for (const item of toCheck) {
      if (item.value) {
        const parsed = parseAnyDate(item.value);
        if (parsed && isPastDateOnly(parsed, nowUtc)) {
          return res.status(400).json({ msg: `${item.name} must be a future date` });
        }
      }
    }
  } catch (e) {
    return res.status(400).json({ msg: 'Invalid date or number format' });
  }

  const db = getDb();
  const vehicles = db.collection('vehicles');

  const existing = await vehicles.findOne({ vehicleNumber });
  if (existing) {
    if (req.file) {
      removeUploadByUrl(`/uploads/${req.file.filename}`);
    }
    return res.status(400).json({ msg: 'Vehicle with this number already exists' });
  }

  const ownerId = req.user.id;
  const rcBookUrl = req.file ? `/uploads/${req.file.filename}` : null;

  const newVehicle = {
    vehicleNumber,
    brand: data.brand,
    model,
    variant: data.variant,
    vehicleType: data.vehicleType,
    fuelType,
    color: data.color,
    manufacturingYear: yearInt,
    registrationDate: data.registrationDate,
    registeredRTO: data.registeredRTO,

    ownerName: data.ownerName,
    phone: data.phone,
    ownershipCount: data.ownershipCount ? parseInt(data.ownershipCount, 10) : 1,
    purchaseDate: data.purchaseDate,
    purchasePrice: data.purchasePrice ? parseFloat(data.purchasePrice) : null,

    insuranceProvider: data.insuranceProvider,
    insuranceExpiry: data.insuranceExpiry,
    pucExpiry: data.pucExpiry,
    rcExpiry: data.rcExpiry,
    roadTaxValidTill: data.roadTaxValidTill,
    fitnessExpiry: data.fitnessExpiry,

    chassisNumber: data.chassisNumber,
    engineNumber: data.engineNumber,
    rcBookUrl,

    currentOdometerKm: data.currentOdometerKm ? parseInt(data.currentOdometerKm, 10) : 0,
    averageMonthlyKm: data.averageMonthlyKm ? parseInt(data.averageMonthlyKm, 10) : 0,

    ownerId,
    createdBy: ownerId,
    role: req.user.role || 'Vehicle Owner',
    verificationStatus: 'Pending',
    isArchived: false,
    createdAt: new Date()
  };

  try {
    await vehicles.insertOne(newVehicle);
    return res.status(201).json({ msg: 'Vehicle added successfully' });
  } catch (e) {
    return res.status(500).json({ msg: 'Error adding vehicle', error: String(e && e.message ? e.message : e) });
  }
});

router.get('/myvehicles', requireAuth, async (req, res) => {
  const db = getDb();
  const vehicles = db.collection('vehicles');

  const cursor = vehicles.find({ ownerId: req.user.id, isArchived: { $ne: true } });
  const results = [];

  for await (const v of cursor) {
    results.push({
      id: String(v._id),
      vehicleNumber: v.vehicleNumber,
      brand: v.brand,
      model: v.model,
      variant: v.variant,
      vehicleType: v.vehicleType,
      year: v.manufacturingYear || v.year,
      fuelType: v.fuelType,
      color: v.color,
      registrationDate: v.registrationDate,
      registeredRTO: v.registeredRTO,
      ownerName: v.ownerName,
      phone: v.phone,
      ownershipCount: v.ownershipCount,
      purchaseDate: v.purchaseDate,
      purchasePrice: v.purchasePrice,
      insuranceProvider: v.insuranceProvider,
      insuranceExpiry: v.insuranceExpiry,
      pucExpiry: v.pucExpiry,
      rcExpiry: v.rcExpiry,
      roadTaxValidTill: v.roadTaxValidTill,
      fitnessExpiry: v.fitnessExpiry,
      chassisNumber: v.chassisNumber,
      engineNumber: v.engineNumber,
      currentOdometerKm: v.currentOdometerKm,
      averageMonthlyKm: v.averageMonthlyKm,
      ownerId: v.ownerId,
      rcBookUrl: v.rcBookUrl,
      verificationStatus: v.verificationStatus || 'Pending',
      isArchived: v.isArchived || false,
      createdAt: v.createdAt ? new Date(v.createdAt).toISOString() : null
    });
  }

  return res.status(200).json(results);
});

router.put('/:vehicle_id', requireAuth, upload.single('rcBook'), async (req, res) => {
  const vehicleId = req.params.vehicle_id;

  let objectId;
  try {
    objectId = new ObjectId(vehicleId);
  } catch (e) {
    return res.status(400).json({ msg: 'Error updating vehicle', error: 'Invalid vehicle ID' });
  }

  const data = req.body || {};
  const updateData = {};

  const directKeys = [
    'vehicleNumber',
    'brand',
    'model',
    'variant',
    'vehicleType',
    'fuelType',
    'color',
    'registrationDate',
    'registeredRTO',
    'ownerName',
    'phone',
    'purchaseDate',
    'insuranceProvider',
    'insuranceExpiry',
    'pucExpiry',
    'rcExpiry',
    'roadTaxValidTill',
    'fitnessExpiry',
    'chassisNumber',
    'engineNumber'
  ];

  for (const key of directKeys) {
    if (Object.prototype.hasOwnProperty.call(data, key)) {
      updateData[key] = data[key];
    }
  }

  if (data.year) updateData.manufacturingYear = parseInt(data.year, 10);
  if (data.ownershipCount) updateData.ownershipCount = parseInt(data.ownershipCount, 10);
  if (data.purchasePrice) updateData.purchasePrice = parseFloat(data.purchasePrice);
  if (data.currentOdometerKm) updateData.currentOdometerKm = parseInt(data.currentOdometerKm, 10);
  if (data.averageMonthlyKm) updateData.averageMonthlyKm = parseInt(data.averageMonthlyKm, 10);

  const db = getDb();
  const vehicles = db.collection('vehicles');

  try {
    const vehicle = await vehicles.findOne({ _id: objectId, ownerId: req.user.id, isArchived: { $ne: true } });
    if (!vehicle) {
      if (req.file) removeUploadByUrl(`/uploads/${req.file.filename}`);
      return res.status(404).json({ msg: 'Vehicle not found or unauthorized' });
    }

    const nextVehicleNumber = updateData.vehicleNumber;
    if (nextVehicleNumber && nextVehicleNumber !== vehicle.vehicleNumber) {
      const conflict = await vehicles.findOne({ vehicleNumber: nextVehicleNumber });
      if (conflict) {
        if (req.file) removeUploadByUrl(`/uploads/${req.file.filename}`);
        return res.status(400).json({ msg: 'Vehicle with this number already exists' });
      }
    }

    if (req.file) {
      if (vehicle.rcBookUrl) {
        removeUploadByUrl(vehicle.rcBookUrl);
      }
      updateData.rcBookUrl = `/uploads/${req.file.filename}`;
    }

    if (Object.keys(updateData).length > 0) {
      await vehicles.updateOne({ _id: objectId }, { $set: updateData });
    }

    return res.status(200).json({ msg: 'Vehicle updated successfully' });
  } catch (e) {
    return res.status(500).json({ msg: 'Error updating vehicle', error: String(e && e.message ? e.message : e) });
  }
});

router.delete('/:vehicle_id', requireAuth, async (req, res) => {
  const vehicleId = req.params.vehicle_id;
  let objectId;
  try {
    objectId = new ObjectId(vehicleId);
  } catch (e) {
    return res.status(500).json({ msg: 'Error archiving vehicle', error: 'Invalid vehicle ID' });
  }

  const db = getDb();
  const vehicles = db.collection('vehicles');
  const services = db.collection('services');

  try {
    const vehicle = await vehicles.findOne({ _id: objectId, ownerId: req.user.id });
    if (!vehicle) {
      return res.status(404).json({ msg: 'Vehicle not found or unauthorized' });
    }

    await services.updateMany({ vehicleId: vehicleId }, { $set: { isArchived: true } });
    await vehicles.updateOne({ _id: objectId }, { $set: { isArchived: true } });

    return res.status(200).json({ msg: 'Vehicle archived successfully' });
  } catch (e) {
    return res.status(500).json({ msg: 'Error archiving vehicle', error: String(e && e.message ? e.message : e) });
  }
});

module.exports = router;
